### PR TITLE
[Security] deprecate multiple providers in context listener

### PR DESCRIPTION
**Pull request by xabbuh**
Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | yes
| Tests pass?   | yes
| Fixed tickets | 
| License       | MIT
| Doc PR        | 

Passing multiple user providers to the context listener does not make
much sense. The listener is only responsible to refresh users for a
particular firewall. Thus, it must only be aware of the user provider
for this particular firewall.

### CODE REVIEW

**Comment by xabbuh** [comment]
This will need a rebase with #21791 being merged up to master.

File: src/Symfony/Component/Security/Http/Firewall/ContextListener.php [hunk 1]
```diff
@@ -44,12 +44,26 @@ class ContextListener implements ListenerInterface
...
+     * @param AuthenticationTrustResolverInterface|null     $trustResolver
+     */
+    public function __construct(TokenStorageInterface $tokenStorage, $userProviders, $contextKey, LoggerInterface $logger = null, EventDispatcherInterface $dispatcher = null, AuthenticationTrustResolverInterface $trustResolver = null)
     {
         if (empty($contextKey)) {
             throw new \InvalidArgumentException('$contextKey must not be empty.');
         }
 
+        if (is_array($userProviders)) {
+            @trigger_error(sprintf('Being able to pass multiple user providers to the constructor of %s is deprecated since version 3.3 and will not be supported anymore in 4.0. Only pass the user provider for the current firewall context instead.', __CLASS__), E_USER_DEPRECATED);
```
Comment (fabpot) [inline src/Symfony/Component/Security/Http/Firewall/ContextListener.php]: missing `"` around `%s`

File: UPGRADE-3.3.md [hunk 2]
```diff
@@ -94,6 +94,9 @@ Process
 Security
 --------
 
+ * Deprecated the ability to pass multiple user providers to the `ContextListener`. Pass only the user provider responsible
+   for the active firewall instead.
```
Comment (fabpot) [inline UPGRADE-3.3.md]: looks like it's not wrapped at 80 chars

**Comment by fabpot** [comment]
Thank you @xabbuh.

**Comment by stof** [comment]
@xabbuh a given firewall can rely on multiple user providers when multiple authentication listeners are enabled (as each of them can use a different user provider)

**Comment by xabbuh** [comment]
will be reverted by #21799

**Pull request merged**